### ISSUE TITLE
GAN Manual Optimization not working for 1.2.0 through master

### ISSUE THREAD

**Issue by import-antigravity** [PyTorchLightning/pytorch-lightning#6282]
## 🐛 Bug

Originally #5545 but since the problem appears to have been fixed and then re-appeared, I thought I'd just create another issue.

## Please reproduce using the BoringModel

https://colab.research.google.com/gist/import-antigravity/0730243bb11b56031110fd6aa7d58971/the-boringmodel.ipynb

### To Reproduce

Alter the `! pip` command at the beginning of the notebook to change the version and observe that after 1.2.0 the GAN does not appear to begin to converge, and just outputs noise. Running the code on version 1.1.8 works.

### Expected behavior

After running the code for 5 epochs it should be clear that the GAN is beginning to create the MNIST digits.

### Environment
```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1+cu101
	- pytorch-lightning: 1.3.0dev
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020
	```

**Comment by ethanwharris** [comment]
Possibly relevant comment from old issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/5545#issuecomment-788321320

**Comment by import-antigravity** [comment]
Update: moving `zero_grad()` calls to *before* any forward passes has fixed the issue. However I want to keep this open as the documentation needs to be updated.

https://pytorch-lightning.readthedocs.io/en/stable/common/optimizers.html?highlight=optimization#manual-optimization

Also, I'm not sure if requiring that zero_grad() is always called first is necessarily a desired behavior.

**Comment by import-antigravity** [comment]
![Screen Shot 2021-03-01 at 5 10 59 PM](https://user-images.githubusercontent.com/24441495/109565899-1b415a80-7ab1-11eb-8b98-67d34030dc51.png)
I'm experimenting with where to put the `zero_grad()` calls, and it seems like it's best to have them separate, before each respective model's forward passes. I would put them all at the beginning but for a GAN you want the gradients for the generator and discriminator to be separate.

**Comment by tchaton** [comment]
Dear @import-antigravity,

From 1.2.0, users are left to properly use call everything.

The `zero_grad` call should be called before each `manual_backward`. As the graph between generator and discriminator are connected.

```
    def training_step(self, batch, batch_idx, optimizer_idx, *args):
        g_opt, d_opt = self.optimizers()

        X, _ = batch

        batch_size = X.shape[0]
        g_X = self.sample_G(batch_size)
        d_x = self.D(X)
        d_z = self.D(g_X)

        # Train generator
        g_loss = self.generator_loss(d_z)
        g_opt.zero_grad()
        self.manual_backward(g_loss, retain_graph=True)
        g_opt.step()

        # Train discriminator
        d_z = self.D(g_X.detach())

        d_loss = self.discriminator_loss(d_x, d_z)
        d_opt.zero_grad()
        self.manual_backward(d_loss)
        d_opt.step()

        self.log_dict({'g_loss': g_loss, 'd_loss': d_loss}, prog_bar=True, logger=True)

``` 

<img width="454" alt="Screenshot 2021-03-02 at 09 46 03" src="https://user-images.githubusercontent.com/12861981/109629782-1c6b9980-7b3c-11eb-9b9a-c07c28b78c97.png">

You can also use `toggle_model` to avoid making errors.

<img width="744" alt="Screenshot 2021-03-02 at 09 57 20" src="https://user-images.githubusercontent.com/12861981/109631262-aff19a00-7b3d-11eb-9496-eefc4fcf7fb1.png">

I will be closing this issue. Re-open if something isn't clear.

Best,
T.C

**Comment by import-antigravity** [comment]
@tchaton,

The code snippet you provided does not work in the notebook I linked.

![Screen Shot 2021-03-02 at 11 24 15 AM](https://user-images.githubusercontent.com/24441495/109679702-d3badd00-7b49-11eb-8ffd-a8611446bdb7.png)

![Screen Shot 2021-03-02 at 11 23 27 AM](https://user-images.githubusercontent.com/24441495/109679671-caca0b80-7b49-11eb-9772-1d333908380b.png)

**Comment by SamuelBroughton** [comment]
> Update: moving `zero_grad()` calls to _before_ any forward passes has fixed the issue. However I want to keep this open as the documentation needs to be updated.
> 
> https://pytorch-lightning.readthedocs.io/en/stable/common/optimizers.html?highlight=optimization#manual-optimization
> 
> Also, I'm not sure if requiring that zero_grad() is always called first is necessarily a desired behavior.

@import-antigravity I also have had this problem recently. I am introducing the Discriminator part way through training.

 I too moved `zero_grad()` before the forward pass which is not reflected in the documentation screen shot you provided @tchaton ?

My training step based on the documentation but altered by moving `zero_grad()` to before:

<img width="894" alt="Screenshot 2021-03-03 at 8 50 25 PM" src="https://user-images.githubusercontent.com/36850153/109808787-82524100-7c62-11eb-87f1-87c6c226e3c6.png">

**Comment by akihironitta** [comment]
@import-antigravity @SamuelBroughton We recently updated the docs for manual optimization, so could you have a look at [the latest docs](https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html)? Please let me know if there's still a problem.

**Comment by import-antigravity** [comment]
> > Update: moving `zero_grad()` calls to _before_ any forward passes has fixed the issue. However I want to keep this open as the documentation needs to be updated.
> > https://pytorch-lightning.readthedocs.io/en/stable/common/optimizers.html?highlight=optimization#manual-optimization
> > Also, I'm not sure if requiring that zero_grad() is always called first is necessarily a desired behavior.
> 
> @import-antigravity I also have had this problem recently. I am introducing the Discriminator part way through training.
> 
> I too moved `zero_grad()` before the forward pass which is not reflected in the documentation screen shot you provided @tchaton ?
> 
> My training step based on the documentation but altered by moving `zero_grad()` to before:
> 
> <img alt="Screenshot 2021-03-03 at 8 57 17 PM" width="869" src="https://user-images.githubusercontent.com/36850153/109809244-0c9aa500-7c63-11eb-8fcc-ca7cfecfb995.png">

Does that code not work? It looks fine.

**Comment by SamuelBroughton** [comment]
> > > Update: moving `zero_grad()` calls to _before_ any forward passes has fixed the issue. However I want to keep this open as the documentation needs to be updated.
> > > https://pytorch-lightning.readthedocs.io/en/stable/common/optimizers.html?highlight=optimization#manual-optimization
> > > Also, I'm not sure if requiring that zero_grad() is always called first is necessarily a desired behavior.
> > 
> > 
> > @import-antigravity I also have had this problem recently. I am introducing the Discriminator part way through training.
> > I too moved `zero_grad()` before the forward pass which is not reflected in the documentation screen shot you provided @tchaton ?
> > My training step based on the documentation but altered by moving `zero_grad()` to before:
> > <img alt="Screenshot 2021-03-03 at 8 57 17 PM" width="869" src="https://user-images.githubusercontent.com/36850153/109809244-0c9aa500-7c63-11eb-8fcc-ca7cfecfb995.png">
> 
> Does that code not work? It looks fine.

Yes this works, I was just supporting what you mentioned about needing to move `zero_grad()` calls.
